Introduce a test framework for the managed group API

kkohbrok opened the pull request:
This PR introduces a framework to make testing the managed group API easier. Note, that I re-oranized the test utils, so `mls_utils` is now the module housing the existing test framework for the `mls_group` API. That is also the reason why the PR looks so big.

I added `Node` to the prelude, which I think is useful, because we export the public tree as `Vec<Option<Node>>`.

I've also added the test frameworks to Codecov's `ignore` list. FWIW, I have two follow-up PRs lined up to do some testing with the test framework, which should test the testing framework somewhat.

Sorry about the mess that are the commits. They'll get squashed anyway before merging to `main`.

franziskuskiefer commented on tests/utils/managed_utils/mod.rs:
```diff
@@ -0,0 +1,583 @@
+    pub(crate) fn store_credentials(
+        &mut self,
+        client_id: Vec<u8>,
+        credential_bundles: Vec<(CiphersuiteName, CredentialBundle)>,
```
This could take a slice, but vec is fine I guess

franziskuskiefer replied:
Yeah, you'd have to consume it.

franziskuskiefer commented on tests/utils/managed_utils/mod.rs:
```diff
@@ -0,0 +1,583 @@
+impl<'ks> KeyStore {
+    pub(crate) fn store_credentials(
+        &mut self,
+        client_id: Vec<u8>,
```
This shuold be a slice.

franziskuskiefer commented on tests/utils/managed_utils/mod.rs:
```diff
@@ -0,0 +1,583 @@
+
+    pub(crate) fn get_credential(
+        &self,
+        client_id: &Vec<u8>,
```
slice

franziskuskiefer commented on tests/utils/managed_utils/mod.rs:
```diff
@@ -0,0 +1,583 @@
+        let index = (OsRng.next_u32() as usize) % self.members.len();
+        // We can unwrap here, because the index is scoped with the size of the
+        // HashSet.
+        let (_, identity) = self.members.iter().nth(index).unwrap().clone();
```
`self.members.get(index)`? Or because we know it must exist `self.members[index]`?

kkohbrok replied:
Right. This is a relic of when `self.members` used to be a `HashMap`.

kkohbrok replied:
Done!

franziskuskiefer commented on tests/utils/managed_utils/mod.rs:
```diff
@@ -0,0 +1,583 @@
+
+        // TODO: Do multiple things.
+        let operation_type = match (OsRng.next_u32() as usize) % 3 {
+            0 => OperationType::Update,
```
If you don't use the enum anywhere else, this looks a little overcomplicated. Why not just match the random number?

kkohbrok replied:
I used to match the random number, but I thought using the enum would make it a bit more readable. If you prefer matching the number, I'm happy to restore that.

kkohbrok replied:
Done!

franziskuskiefer commented on tests/utils/managed_utils/mod.rs:
```diff
@@ -0,0 +1,583 @@
+        let (messages, welcome_option) = match action_type {
+            ActionType::Commit => {
+                // Let the function generate the key package bundle.
+                member_group_state.self_update(None)?
```
So no way to test with an externally generated key package bundle?

kkohbrok replied:
Good point, we should probably allow that.

kkohbrok replied:
Done!

franziskuskiefer commented on tests/utils/managed_utils/mod.rs:
```diff
@@ -0,0 +1,583 @@
+
+pub struct ManagedTestSetup<'client_lifetime> {
+    pub number_of_clients: usize,
+    // The clients identity is its position in the vector in be_bytes.
```
😬 I don't think I like this approach for identities. But it's as good or bad as any other thing we can do here.
We shoudl have a function to get the identity for a client though, i.e. don't duplicate `i.to_be_bytes().to_vec()`.

kkohbrok replied:
Feel free to suggest another approach. I'm not particularly attached to this one, so I'm happy to replace it with something else. This just seemed convenient to implement.

franziskuskiefer commented on tests/utils/managed_utils/mod.rs:
```diff
@@ -0,0 +1,583 @@
+        // Create credentials first to avoid borrowing issues.
+        for i in 0..number_of_clients {
+            let identity = i.to_be_bytes().to_vec();
+            // For now, everyone supports all ciphersuites.
```
Another thing to extend later.

kkohbrok replied:
Yes, I'll add an issue for that once this PR lands.

franziskuskiefer commented on tests/utils/managed_utils/mod.rs:
```diff
@@ -0,0 +1,583 @@
+    }
+
+    /// Deliver a Welcome message to group `group` to the intended recipients.
+    pub fn deliver_welcome(&self, welcome: Welcome, group: &Group) -> Result<(), SetupError> {
```
You pass in the `group` here and pull information from it. If that's intended in you test setup, it needs description of how it is supposed to be used and how it allows testing things. In the real world the `group_config` must be extracted from the `Welcome` and the `public_tree` might be in there as well, or delivered separately. (All things that are impossible to test with this function.)

kkohbrok replied:
Fair point. I guess I wrote this function mostly with `pub(crate)` use in mind, but it's certainly something we will want to use outside. I should probably switch from passing in `&Group` everywhere to passing in `GroupId`.

kkohbrok replied:
Made the documentation more precise.

franziskuskiefer commented on tests/utils/managed_utils/mod.rs:
```diff
@@ -0,0 +1,583 @@
+        drop(sender_group);
+        drop(sender_groups);
+        println!("Group members after distribution:");
+        // Check that the group states of all members match.
```
Why? I can see that this is a nice thing to check in many cases. But there are at least as many interesting cases where this really shouldn't be the case.

kkohbrok replied:
Fair point. Maybe it would be worth passing in a switch that enables/disables this behaviour or moving it to an external function of `Group`.

kkohbrok replied:
Moved it to an external function that can be called independently.

franziskuskiefer commented on tests/utils/managed_utils/mod.rs:
```diff
@@ -0,0 +1,583 @@
+        for (index, m_id) in &group.members {
+            println!("Id: {:?}, Index: {:?}", m_id, index);
+            let m = clients.get(m_id).unwrap().borrow();
+            let group_states = m.groups.borrow_mut();
```
Why do you need this mutably?

kkohbrok replied:
Done!

franziskuskiefer commented on tests/utils/managed_utils/mod.rs:
```diff
@@ -0,0 +1,583 @@
+        }
+        let mut new_member_ids = Vec::new();
+        for _ in 0..number_of_members {
+            let new_member_id = clients
```
Ugh... So you want to move some of the closures here out to functions with names?

kkohbrok replied:
Right, this is very ugly, but I didn't really know how to do it better. I think moving out the closures won't do that much. Maybe it would be worth using `HashSet`s here?

kkohbrok replied:
Done!

franziskuskiefer commented on tests/utils/managed_utils/mod.rs:
```diff
@@ -0,0 +1,583 @@
+                .unwrap();
+            new_member_ids.push(new_member_id.clone());
+        }
+        drop(clients);
```
Why do you need this? 🤔

kkohbrok replied:
Removed!

franziskuskiefer commented on tests/utils/managed_utils/mod.rs:
```diff
@@ -0,0 +1,583 @@
+                            .find(|&member_id| member_id == client_id)
+                            .is_none())
+                })
+                .unwrap();
```
So this panics if there's no client we could add?

kkohbrok replied:
Well, there really should be one. We check in the beginning if enough clients exist.

franziskuskiefer reviewed (changes requested):
I skipped some parts in `managed_utils/mod.rs` because I'm not sure how all this is
supposed to work.

We need at least one very basic example here that uses the test framework to
know how you intend it to be used. (You can add doc tests if you don't want to
add "real" ones.)
A description of how it works and how it enables test coverage would be also nice.

You generate random numbers and then compute them modulo to get something in the
desired range. This will get you skewed results. That's not super important here
but it will certainly lead to some things being tested more than others.

Some things can't be (nicely) tested with this. I made comments where I see issues
(I'm sure there are more). We don't have to fix all of them but file a follow
up issue with everything we know would be good to have in here but is currently
missing.

All the `drop`s make me think that the code is probably not ideal. But we can
refactor this later.

kkohbrok commented on tests/utils/mls_utils/mod.rs:
```diff
@@ -0,0 +1,297 @@
+    get_random_vec(n)
+}
+
+// Not currently used.
```
I only moved the `mls_utils`part and did not otherwise touch it. I'd rather not change anything there.

franziskuskiefer replied:
If you don't change anything in here, please move it to a separate PR.

kkohbrok replied:
Done!

kkohbrok commented on tests/utils/mls_utils/mod.rs:
```diff
@@ -0,0 +1,297 @@
+        test_clients.insert(client.name, RefCell::new(test_client));
+    }
+    // Initialize all of the groups, each group gets assigned a sequential group
+    // id. TODO: Depending on the use case, it might be hard to figure out which
```
See above. I only moved the `mls_utils` code and would rather not touch it any further in this PR.

kkohbrok commented on src/prelude.rs:
```diff
@@ -25,6 +25,7 @@ pub use crate::messages::{
     proposals::{AddProposal, RemoveProposal, UpdateProposal},
     Welcome,
 };
+pub use crate::tree::node;
```
I can also put it in `lib.rs`. But I think the DS will want access to this. Not only for indices, but also, for example, to pull credentials from the public tree.

kkohbrok replied:
Done!

kkohbrok commented on tests/utils/managed_utils/mod.rs:
```diff
@@ -0,0 +1,583 @@
+#![allow(dead_code)]
+/// We allow dead code here due to the following issue:
+/// https://github.com/rust-lang/rust/issues/46379, which would otherwise create
```
When I remove this and I run `cargo test`, I get a ton of `function is never used` warnings for functions that are used either in other functions of the framework or in functions of existing tests (that are not (yet) part of this PR). I think that's related to the issue I linked, but may have other causes. Can you do me a favor and test this on your machine? It could also just be a local issue.

kkohbrok commented:
Thanks for the thorough review! You raise some very good points.

> I skipped some parts in `managed_utils/mod.rs` because I'm not sure how all this is
> supposed to work.

I have more documentation on the way. Sorry about that.

> We need at least one very basic example here that uses the test framework to
> know how you intend it to be used. (You can add doc tests if you don't want to
> add "real" ones.)
> A description of how it works and how it enables test coverage would be also nice.

No problem. I have a few lined up already and will add one of them to the PR.

> You generate random numbers and then compute them modulo to get something in the
> desired range. This will get you skewed results. That's not super important here
> but it will certainly lead to some things being tested more than others.

Good point. I agree that this is pretty ad-hoc. Is there a more way standard way of doing this?

> Some things can't be (nicely) tested with this. I made comments where I see issues
> (I'm sure there are more). We don't have to fix all of them but file a follow
> up issue with everything we know would be good to have in here but is currently
> missing.

Another good point. So far, I've mostly geared the framework to enable testing if managed group works in cases it should work. I agree that there should be room for more direct manipulation and testing failure cases as well.

> All the `drop`s make me think that the code is probably not ideal. But we can
> refactor this later.

I'm also very unhappy with all the drops, but from how I understand it, dealing with the lifetimes necessitates the `RefCell`s, which in turn necessitate dropping state every now and again. I'll try to see if I can get rid of some more of them.

kkohbrok commented on tests/utils/mls_utils/mod.rs:
```diff
@@ -0,0 +1,297 @@
+#![allow(dead_code)]
```
The way I understand https://github.com/rust-lang/rust/issues/46379, non-test modules in `tests` get compiled individually with each test. At least on my machine, this means I get warnings, because not every test uses every function of the framework and reports the non-used functions as dead code. Not sure if there is a really good solution to this.

kkohbrok commented on tests/utils/mod.rs:
```diff
@@ -1,296 +1,2 @@
-    };
-    let _test_setup = setup(test_setup_config);
-}
+pub mod managed_utils;
```
Done!

kkohbrok commented on tests/utils/managed_utils/client.rs:
```diff
@@ -0,0 +1,142 @@
+use std::{cell::RefCell, collections::HashMap};
```
Done!

kkohbrok commented on tests/utils/managed_utils/client.rs:
```diff
@@ -0,0 +1,142 @@
+    /// Name of the client.
+    pub(crate) identity: Vec<u8>,
+    /// Ciphersuites supported by the client.
+    pub(crate) _ciphersuites: Vec<CiphersuiteName>,
```
Done!

kkohbrok commented on tests/utils/managed_utils/client.rs:
```diff
@@ -0,0 +1,142 @@
+    pub(crate) identity: Vec<u8>,
+    /// Ciphersuites supported by the client.
+    pub(crate) _ciphersuites: Vec<CiphersuiteName>,
+    pub(crate) key_store: &'managed_group_lifetime KeyStore,
```
Done!

kkohbrok commented on tests/utils/managed_utils/client.rs:
```diff
@@ -0,0 +1,142 @@
+}
+
+impl<'managed_group_lifetime> Client<'managed_group_lifetime> {
+    pub fn get_fresh_key_package(
```
Done!

kkohbrok commented on tests/utils/managed_utils/client.rs:
```diff
@@ -0,0 +1,142 @@
+            .ok_or(ClientError::CiphersuiteNotSupported)?;
+        let mandatory_extensions = Vec::new();
+        let key_package_bundle: KeyPackageBundle = KeyPackageBundle::new(
+            &[ciphersuite.name()],
```
Done! Now actually uses the first ciphersuite in the list and just announces the rest.

kkohbrok commented on tests/utils/managed_utils/client.rs:
```diff
@@ -0,0 +1,142 @@
+        group_id: GroupId,
+        managed_group_config: ManagedGroupConfig,
+        ciphersuite: &Ciphersuite,
+    ) -> Result<Vec<Option<Node>>, ClientError> {
```
Yes, that didn't really make sense. Fixed!

kkohbrok commented on tests/utils/managed_utils/default_callbacks.rs:
```diff
@@ -0,0 +1,57 @@
+        }
+    }
+}
+fn error_occured(managed_group: &ManagedGroup, error: ManagedGroupError) {
```
Done!

kkohbrok commented on tests/utils/managed_utils/mod.rs:
```diff
@@ -0,0 +1,583 @@
+#[derive(Debug)]
+pub struct KeyStore {
+    // Maps a client Id and a ciphersuite to a CredentialBundle.
+    credential_bundles: HashMap<(Vec<u8>, CiphersuiteName), CredentialBundle>,
```
Done!

kkohbrok commented:
I made a few passes and (I think I) addressed all reviewer comments. In particular, I added a lot of documentation and changed the structure to get rid of quite a few `drop`s. I also added a brief example that used the framework. Bigger tests using the framework will follow along with a few bugfixes 🙂

franziskuskiefer commented on tests/test_managed_api.rs:
```diff
@@ -0,0 +1,38 @@
+        let (_, target_id) = group.members[3].clone();
+        setup
+            .remove_clients(ActionType::Commit, group, &remover_id, vec![target_id])
+            .unwrap();
```
Ok so this is how I use the framework to create a group and add/remove members to/from a group. But how do I test something. I guess we always want to verify the group state after an operation or at the end. Could you do this here so we know how to then write other tests?

franziskuskiefer commented on tests/utils/managed_utils/client.rs:
```diff
@@ -108,18 +123,26 @@ impl<'managed_group_lifetime> Client<'managed_group_lifetime> {
-        Ok(group_state.process_messages(messages)?)
+        for message in messages {
+            let group_id = GroupId {
+                value: message.group_id(),
```
This will not be exposed in future (when we finally cleaned up the API). You can use `GroupId::from_slice`. Or you make your `self.groups` use the raw byte vector instead of the `GroupId`.

franziskuskiefer commented on tests/utils/managed_utils/client.rs:
```diff
@@ -139,4 +162,73 @@ impl<'managed_group_lifetime> Client<'managed_group_lifetime> {
+        &self,
+        action_type: ActionType,
+        group_id: &GroupId,
+        target_indices: &[usize],
```
🙈

franziskuskiefer commented on tests/utils/managed_utils/mod.rs:
```diff
@@ -232,26 +263,34 @@ impl<'ks> ManagedTestSetup<'ks> {
-    /// with the `add_members` function.
+    /// with the `add_members` function. If not enough clients are left that are
+    /// not already members of this group, this function will return an error.
+    /// TODO: Make this function ensure that the given members support the
```
Add issue number or something.

kkohbrok replied:
Done! #310.

franziskuskiefer reviewed (changes requested):
A couple more things.

The easiest thing for generating random numbers in a range is [gen_range](https://docs.rs/rand/0.8.2/rand/trait.Rng.html#method.gen_range).

I added a comment to #219 with some of the things we should add to the framework later.

kkohbrok reviewed (commented):
I made another pass and (hopefully) addressed all comments.

franziskuskiefer commented on tests/utils/managed_utils/client.rs:
```diff
@@ -0,0 +1,142 @@
+        self.key_package_bundles
+            .borrow_mut()
+            .insert(key_package_bundle.key_package().hash(), key_package_bundle);
+        Ok(key_package)
```
You already have the lifetimes and such. So should work. But lets keep it like this for now.

franziskuskiefer reviewed (approved):
Ok let's get it in! 🙂
